adr/0009-extension-structure.adoc

```diff
@@ -0,0 +1,280 @@
+* The `runtime` module should focus on the internal implementation and runtime logic, delegating public APIs to the `runtime-api` module.
+* The `runtime-api` module contains the public API  allowing other extensions to depend on it without pulling in the full extension.
+
+For extensions requiring tight-coupling, the `runtime` module can still be used to expose public APIs, but this should be avoided when possible.
+
+IMPORTANT: Once an extension has a `runtime-api` module, it should not expose public APIs in the `runtime` module. This avoids confusion and ensures that consumers can clearly distinguish between public APIs and internal implementation details. Also, it would not be possible to move the public API back into the `runtime` module.
+
+[cols="1,3,2",options="header"]
+|===
+| Scenario | Recommended Structure | Notes
+
+| Small/simple extension not reused by others
+| `runtime`
+| Keep everything in `runtime`. Avoid unnecessary modularity.
+
+| Extension exposes public types used by application code or other extensions
+| `runtime-api` + `runtime`
+| Split APIs (annotations, interfaces, utility classes) into `runtime-api`. Keep internal logic in `runtime`.
+
+| Extension contributes dev-mode-only logic (e.g., Dev UI)
+| `runtime` + `runtime-dev`
+| Add `runtime-dev` to isolate dev-only classes. Avoid shipping to production.
+
+| Extension defines types meant to be implemented by others (e.g., customizers, listeners)
+| `runtime-api` + `runtime`
+| Consider these types part of the SPI. Place them in `runtime-api` to allow loose coupling.
+
+| Extension wants to enforce tight coupling (full extension must be present)
+| `runtime` only
+| Expose public types directly from `runtime`. Use cautiously; limits flexibility and reuse.
+
+| Extension depends on or provides code generation
+| `runtime` + `codegen`
+| Keep codegen logic isolated. Optional, depending on feature set.
+|===
+
+=== Package Name Rules
+
+Extensions must use a well-defined package structure to avoid split packages.
+
+[NOTE]
+====
+In this section, rules are given for the `io.quarkus` and `io.quarkiverse` namespaces.
+When the root is different, the rules apply similarly, replacing `io.quarkus` or `io.quarkiverse` with the appropriate root package.
+For example: `org.apache.camel.quarkus.component` would use `org.apache.camel.quarkus.component.runtime`, `org.apache.camel.quarkus.component.dev`, etc.
+====
+
+==== Root package name
+
+To transform an extension name into a package name, use the following pattern:
+
+* `io.quarkus.<extension-name>` for Quarkus core extensions.
+* `io.quarkiverse.<extension-name>` for Quarkiverse extensions.
+* If the extension name contains a hyphen, it is recommended to replace it with an underscore (e.g., `quarkus-foo-bar` becomes `io.quarkus.foo_bar`). 
+
+==== `runtime` module
+* `io.<quarkus|quarkiverse>.<extension-name>.runtime.internal|impl`: Internal implementation. Not part of the public API.
```

Comment:
Should we decide what to do here before finalizing the ADR?

Reply:
I recommended "internal".